apps/prairielearn/src/ee/auth/lti13/lti13Auth.ts

```diff
@@ -0,0 +1,259 @@
+  'https://purl.imsglobal.org/spec/lti/claim/deployment_id': z.string(),
+  'https://purl.imsglobal.org/spec/lti/claim/target_link_uri': z.string(),
+  'https://purl.imsglobal.org/spec/lti/claim/resource_link': z.object({
+    id: z.string(),
+    description: z.string().nullable(),
+    title: z.string().nullable(),
+  }),
+  sub: z.string(),
+  'https://purl.imsglobal.org/spec/lti/claim/roles': z.string().array(),
+});
+
+//
+// Helper functions
+//
+
+async function authenticate(req: Request, res: Response): Promise<any> {
+  const myPassport = await setupPassport(req.params.lti13_instance_id);
+  return new Promise((resolve, reject) => {
+    // https://www.imsglobal.org/spec/security/v1p0/#step-3-authentication-response
+    OIDCAuthResponseSchema.parse(req.body);
+
+    // Callback arguments described at
+    // https://github.com/jaredhanson/passport/blob/33b92f96616642864844753a481df7c5b823e047/lib/middleware/authenticate.js#L34
+    myPassport.authenticate(`lti13`, ((err, user, info) => {
+      if (err) {
+        reject(err);
+      } else if (!user) {
+        // The authentication libraries under openid-connect will fail (silently) if the key length
+        // is too small, like with the Canvas development keys. It triggers that error in PL here.
+        reject(
+          error.make(400, `Authentication failed, before user validation.`, {
+            err,
+            user,
+            info_raw: info,
+            info: info?.toString(),
+          }),
+        );
+      } else {
+        resolve(user);
+      }
+    }) as passport.AuthenticateCallback)(req, res);
+  });
+}
+
+async function launchFlow(req: Request, res: Response, next: NextFunction) {
+  // https://www.imsglobal.org/spec/security/v1p0/#step-1-third-party-initiated-login
+
+  const parameters = { ...req.body, ...req.query };
+
+  OIDCLaunchFlowSchema.parse(parameters);
+
+  const myPassport = await setupPassport(req.params.lti13_instance_id);
+  myPassport.authenticate('lti13', {
+    response_type: 'id_token',
+    lti_message_hint: parameters.lti_message_hint,
+    login_hint: parameters.login_hint,
+    prompt: 'none',
+    response_mode: 'form_post',
+    failWithError: true,
+  } as passport.AuthenticateOptions)(req, res, next);
+}
+
+async function setupPassport(lti13_instance_id: string) {
+  const lti13_instance = await selectLti13Instance(lti13_instance_id);
+
+  const localPassport = new passport.Passport();
+  const issuer = new Issuer(lti13_instance.issuer_params);
+  const client = new issuer.Client(lti13_instance.client_params, lti13_instance.keystore);
+
+  localPassport.use(
+    'lti13',
+    new Strategy(
+      {
+        client: client,
+        passReqToCallback: true,
+      },
+      // Passport verify function
+      callbackify(verifyAsync),
+    ),
+  );
+
+  return localPassport;
+}
+
+async function verifyAsync(req: Request, tokenSet: TokenSet) {
+  const lti13_claims = tokenSet.claims();
+
+  LTI13Schema.parse(lti13_claims);
```

Comment:
Nit ([parse, don't validate](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/) - this doesn't make a difference now with the current schema, but it's the conventional way to do things).

```suggestion
  const lti13_claims = LTI13Schema.parse(tokenSet.claims());
```

Reply:
https://www.imsglobal.org/spec/lti/v1p3#user-identity-claims

Reply:
You can use [`.passthrough()`](https://zod.dev/?id=passthrough) on the schema to allow other unknown keys through.

Independent of that, I'd recommend including `nonce` in the schema here; it looks like that's required below.